client/verta/verta/pipeline/_registered_pipeline.py

```diff
@@ -0,0 +1,174 @@
+
+    def _to_pipeline_definition(self) -> Dict[str, Any]:
+        """
+        Create a complete pipeline definition dict from a name and PipelineGraph.
+        Used in conjunction with the client function for creating a registered
+        pipeline from a pipeline graph.
+        """
+        return {
+            "pipeline_version_id": self.id,
+            "graph": self.pipeline_graph._to_graph_definition(),
+            "predecessors": self.pipeline_graph._to_steps_definition(),
+        }
+
+    def _to_pipeline_configuration(
+        self, pipeline_resources: Optional[Dict[str, Resources]] = None
+    ) -> Dict[str, Any]:
+        """
+        Build a pipeline configuration dict for this pipeline. The
+        `env` and `build` keys are not included in the configuration
+        resulting in default values being used by the backend.
+
+        Parameters
+        ----------
+        pipeline_resources : Resources
+
+        Returns
+        -------
+        Dictionary representation of a pipeline configuration.
+        """
+        steps = list()
+        for step in self.pipeline_graph.steps:
+            step_config = {
+                "name": step.name,
+            }
+            if pipeline_resources:
+                step_res = pipeline_resources.get(step.name, None)
+                if step_res:
+                    step_config["resources"] = pipeline_resources.pop(
+                        step.name
+                    )._as_dict()
+            steps.append(step_config)
+        if pipeline_resources:
+            raise ValueError(
+                f"pipeline_resources contains resources for steps not in "
+                f"the pipeline {pipeline_resources.keys()}"
+            )
+        return {
+            "pipeline_version_id": self.id,
+            "steps": steps,
+        }
+
+    @classmethod
+    def _from_pipeline_definition(
+        cls,
+        registered_model_version: RegisteredModelVersion,
+        conn: Connection,
+        conf: Configuration,
+    ) -> "RegisteredPipeline":
+        """
+        Create a Pipeline instance from a specification dict. Used when
+        fetching a registered pipeline from the Verta backend.
+
+        Parameters
+        ----------
+        registered_model_version : :class:`~verta.registry.entities.RegisteredModelVersion`
+            RegisteredModelVersion object associated with this pipeline.
+        pipeline_definition : dict
+            Specification dict from which to create the Pipeline.
+        conn : :class:`~verta._internal_utils._utils.Connection`
+            Connection object for fetching the models and model versions associated with steps.
+        conf : :class:`~verta._internal_utils._utils.Configuration`
+            Configuration object for fetching the models and model versions associated with steps.
+        """
+        pipeline_definition_str = registered_model_version.get_artifact(
+            "pipeline.json"
+        ).read()
```

Comment:
Have you tested this? I would assume `get_artifact()` returns either a `str` or a `dict`, so I think `read()` here would throw an error.

Reply:
artifact handling has been refactored and tested